service/controller/internal/detection/detection.go

```diff
@@ -72,21 +72,60 @@ func (d *Detection) ShouldUpdate(ctx context.Context, cr v1alpha1.AWSConfig) (bo
 		return false, microerror.Mask(err)
 	}
 
-	if cc.Status.TenantCluster.MasterInstance.Type != key.MasterInstanceType(cr) {
-		d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("detected the tenant cluster should update due to master instance type changes: cc.Status.TenantCluster.MasterInstance.Type is %q while key.MasterInstanceType(cr) is %q", cc.Status.TenantCluster.MasterInstance.Type, key.MasterInstanceType(cr)))
-		return true, nil
-	}
-	if cc.Status.TenantCluster.WorkerInstance.DockerVolumeSizeGB != key.WorkerDockerVolumeSizeGB(cr) {
-		d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("detected the tenant cluster should update due to worker instance docker volume size changes: cc.Status.TenantCluster.WorkerInstance.DockerVolumeSizeGB is %q while key.WorkerDockerVolumeSizeGB(cr) is %q", cc.Status.TenantCluster.WorkerInstance.DockerVolumeSizeGB, key.WorkerDockerVolumeSizeGB(cr)))
-		return true, nil
+	imageID, err := key.ImageID(cr, cc.Spec.TenantCluster.Release)
+	if err != nil {
+		return false, microerror.Mask(err)
 	}
-	if cc.Status.TenantCluster.WorkerInstance.Type != key.WorkerInstanceType(cr) {
-		d.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("detected the tenant cluster should update due to worker instance type changes: cc.Status.TenantCluster.WorkerInstance.Type is %q while key.WorkerInstanceType(cr) is %q", cc.Status.TenantCluster.WorkerInstance.Type, key.WorkerInstanceType(cr)))
-		return true, nil
+
+	changeCandidates := []struct {
```

Comment:
This was getting unwieldy so I turned it into a slice. The old detection included `MasterInstanceType`, `WorkerDockerVolumeSizeGB`, `WorkerInstanceType`, and `VersionBundleVersion`. I dropped `VersionBundleVersion` since we don't want to force an upgrade when the ignition hasn't changed and added mater and worker ignition hash and image (AMI).